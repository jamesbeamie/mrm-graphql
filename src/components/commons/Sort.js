import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Dropdown from './Dropdown';
import '../../assets/styles/sort.scss';

class Sort extends Component {
  state = {
    optionName: null,
  };

  componentWillReceiveProps() {
    // eslint-disable-next-line no-unused-expressions
    this.props.hideDropdownMenu && this.Dropdown.current.toggleVisibility;
  }

  Dropdown = React.createRef();

  /**
   * It toggles the visibility of the options children
   *
   * @param {Object} optionName
   *
   * @returns {void}
   */
  toggleOptions = optionName => () => {
    this.setState({ optionName });
  };

  /**
   * It renders the options to sort by
   *
   * @param {Object} obj
   *
   * @returns {JSX}
   */
  renderObjectOptions = options =>
    Object.keys(options).map(option => (
      // eslint-disable-next-line
      <span
        key={option}
        className={
          this.state.optionName === option
            ? 'filter-options__disable'
            : 'filter-options'
        }
        onClick={this.toggleOptions(option)}
      >
        {option}
        {this.state.optionName === option && (
          <div className="filter-options__children">
            {this.renderOptionsChildren(options[option], option)}
          </div>
        )}
      </span>
    ));

  /**
   * It renders the options children
   *
   * @param {Object} subOptions
   * @param {string} optionName
   *
   * @returns {JSX}
   */
  renderOptionsChildren = (subOptions, optionName) => (
    <React.Fragment>
      {subOptions.map(option => (
        // eslint-disable-next-line
        <span
          key={option.id}
          className="filter-options__children-list"
          onClick={this.props.fetchSortedData(optionName, option.id)}
        >
          {option[Object.keys(option)[1]]}
        </span>
      ))}
    </React.Fragment>
  );

  /**
   * It renders the sort options which is an array
   *
   * @param {array} options
   *
   * @returns {JSX}
   */
  renderArrayOptions = () => <div />; // eslint-disable-line

  render() {
    const { sortOptions, withChildren } = this.props;
    return (
      <div className="filter-options-container">
        <div className="container-inner">
          <span>Sort by:</span>
          <Dropdown
            icon="&#x21F5;"
            content={
              withChildren
                ? this.renderObjectOptions(sortOptions)
                : this.renderArrayOptions(sortOptions)
            }
            ref={this.Dropdown}
          />
        </div>
      </div>
    );
  }
}

Sort.propTypes = {
  sortOptions: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
    .isRequired,
  withChildren: PropTypes.bool,
  fetchSortedData: PropTypes.func.isRequired,
  hideDropdownMenu: PropTypes.bool,
};

Sort.defaultProps = {
  withChildren: false,
  hideDropdownMenu: false,
};

export default Sort;
