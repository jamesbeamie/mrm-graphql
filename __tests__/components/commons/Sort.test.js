import React from 'react';
import { mount } from 'enzyme';
import Sort from '../../../src/components/commons/Sort';

describe('Sort Component', () => {
  const props = {
    fetchSortedData: jest.fn(),
    sortOptions: {
      role: [
        {
          id: '1',
          name: 'admin',
        },
        {
          id: '2',
          name: 'user',
        },
      ],
    },
    withChildren: true,
    hideDropdownMenu: false,
  };

  const wrapper = mount(<Sort {...props} />);

  it('should render component without errors', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should display the option in a dropdown menu component', () => {
    wrapper.find('.dropdown-caret button').simulate('mousedown', {
      type: 'focus',
    });
    expect(wrapper.find('.filter-options').length).toEqual(1);
    expect(wrapper.find('.filter-options').text()).toEqual('role');
  });

  const newWrapper = mount(<Sort {...{ ...props, withChildren: false, hideDropdownMenu: true }} />);
  it('should not display the option in a dropdown menu component', () => {
    newWrapper.find('.dropdown-caret button').simulate('mousedown', {
      type: 'focus',
    });
    expect(newWrapper.find('.filter-options').length).toEqual(0);
  });
});
