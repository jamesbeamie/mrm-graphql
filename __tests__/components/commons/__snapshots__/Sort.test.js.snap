// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Component should render component without errors 1`] = `
<Sort
  fetchSortedData={[MockFunction]}
  hideDropdownMenu={false}
  sortOptions={
    Object {
      "role": Array [
        Object {
          "id": "1",
          "name": "admin",
        },
        Object {
          "id": "2",
          "name": "user",
        },
      ],
    }
  }
  withChildren={true}
>
  <div
    className="filter-options-container"
  >
    <div
      className="container-inner"
    >
      <span>
        Sort by:
      </span>
      <Dropdown
        content={
          Array [
            <span
              className="filter-options"
              onClick={[Function]}
            >
              role
            </span>,
          ]
        }
        icon="⇵"
      >
        <div
          className="dropdown-caret"
        >
          <button
            onMouseDown={[Function]}
          >
            ⇵
          </button>
        </div>
      </Dropdown>
    </div>
  </div>
</Sort>
`;
